packages/nimble-blazor/NimbleBlazor/Components/NimbleTable.razor.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System.Text.Json;
+using Microsoft.AspNetCore.Components;
+using Microsoft.JSInterop;
+
+namespace NimbleBlazor;
+
+/// <summary>
+/// A table component
+/// </summary>
+/// <typeparam name="TData">Represents the type for a row of data in the table (an element of the <see cref="Data"/>).</typeparam>
+/// <remarks>The type represented by <see cref="TData"/> should not have any hierarchy. All aspects that can be serialized
+/// should be at the top-level.</remarks>
+public partial class NimbleTable<TData> : ComponentBase
+{
+    private ElementReference _table;
+    private bool _shouldRender = false;
+    private IEnumerable<TData> _data = Enumerable.Empty<TData>();
+    internal static string SetTableDataMethodName = "NimbleBlazor.Table.setData";
+
+    [Inject]
+    private IJSRuntime? JSRuntime { get; set; }
+
+    [Parameter]
+    public RenderFragment? ChildContent { get; set; }
+
+    /// <summary>
+    /// Gets or sets the data for the table.
+    /// </summary>
+    [Parameter]
+    public IEnumerable<TData> Data
+    {
+        get
+        {
+            return _data;
+        }
+        set
+        {
+            _data = value;
+            _shouldRender = true;
+        }
+    }
+
+    /// <summary>
+    /// Gets or sets a callback that's invoked when the data changes
+    /// </summary>
+    [Parameter]
+    public EventCallback<IEnumerable<TData>> DataChanged { get; set; }
+
+    [Parameter(CaptureUnmatchedValues = true)]
+    public IDictionary<string, object>? AdditionalAttributes { get; set; }
+
+    /// <inheritdoc/>
+    /// <exception cref="JsonException"></exception>
+    protected override Task OnAfterRenderAsync(bool firstRender)
+    {
+        var options = new JsonSerializerOptions { MaxDepth = 3 };
+        _shouldRender = false;
+        return JSRuntime!.InvokeVoidAsync(SetTableDataMethodName, JsonSerializer.Serialize(Data, options), _table).AsTask();
```

Comment:
> The Data property is serialized using JsonSerializer and passed to the nimble-table web component in AfterRenderAsync. This is because the only valid place to do JSInterop.InvokeAsync(...) calls is within AfterRenderAsync.

- [ ] Can you include some context for this in a comment here? I vaguely remember there being some investigation to get to calling it in this method, is their any insight to capture / docs to link to?
- [ ] We call InvokeVoidAsync in the Show and Close methods for the drawer and dialog. Are those calls incorrect?
- [ ] Alternatively / additionally, is there some Contributing documentation we should capture for Nimble Blazor devs around JSInterop and the right time to call things?

Reply:
Yeah I don't think that PR comment is accurate. The [Blazor docs for JS Interop](https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-javascript-from-dotnet?view=aspnetcore-7.0#component-razor-example-invokeasync) show JS being invoked in response to a button click. So I think we could invoke JS in Data.set in at least some cases (I'm still doing some testing locally)

Reply:
This also seems to work, for example (without overriding `ShouldRender()`):
```cs
    /// <summary>
    /// Gets or sets the data for the table.
    /// </summary>
    [Parameter]
    public IEnumerable<TData> Data
    {
        get
        {
            return _data;
        }
        set
        {
            _data = value;
            _ = SendTableDataToJavascriptAsync();
        }
    }

    /// <inheritdoc/>
    /// <exception cref="JsonException"></exception>
    protected async override Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            await SendTableDataToJavascriptAsync();
        }
    }

    private Task SendTableDataToJavascriptAsync()
    {
        var options = new JsonSerializerOptions { MaxDepth = 3 };
        return JSRuntime!.InvokeVoidAsync(SetTableDataMethodName, JsonSerializer.Serialize(Data, options), _table).AsTask();
    }
```

Reply:
My main hesitation to this approach was the act of invoking an async method in a non-async context, but perhaps this is just unnecessary caution on my part. The question, to me, is what is wrong with the approach of just doing the call in `AfterRenderAsync`? Is it for cases where we might be unnecessarily rendering the data because of some state change that doesn't require the data to re-render?

Reply:
I think the code that you have covers the normal cases. The variant I pasted would allow getting updated table data from C# to JS independently of the Blazor rendering paths (since setting the Data property sends it to JS directly) but I'm not sure that's desirable anyway.

Reply:
@rajsite @msmithNI, I don't mind going with Malcolm's approach, but are we comfortable with having something in place that looks "hackish" to avoid a code analysis flag (caching the return value of the async function in an unused local)?

Reply:
I don't think I feel strongly about either approach.

MS docs say ["The OnAfterRender(Boolean) and OnAfterRenderAsync(Boolean) lifecycle methods are useful for performing interop"](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.componentbase.onafterrenderasync?view=aspnetcore-7.0#remarks) so maybe that's enough to go with OnAfterRenderAsync `¯\_(ツ)_/¯`

Is one approach better than another from a debugging perspective? I'd imagine is an error was thrown, ie during JSON parse in JS, that OnAfterRenderAsync may be able to bubble up an exception better since it actually waits on the response?

> Is it for cases where we might be unnecessarily rendering the data because of some state change that doesn't require the data to re-render?

I think your _shouldRender flag mitigates that concern for me. I guess my other concern is that is there documentation about when OnAfterRenderAsync is called? That doc page doesn't make it clear that, for example, all property updates result in a call to OnAfterRenderAsync

Reply:
> I guess my other concern is that is there documentation about when OnAfterRenderAsync is called? That doc page doesn't make it clear that, for example, all property updates result in a call to OnAfterRenderAsync

The [Rendering conventions for ComponentBase](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/rendering?view=aspnetcore-7.0#rendering-conventions-for-componentbase) topic covers what triggers render operations. From that it actually doesn't seem right to only render on Data property updates and maybe we should not override ShouldUpdate at all.

Then in OnAfterRenderAsync we should change the _shouldUpdate flag to something like _dataUpdated and only do the serialization and data sending when that flag is set. Maybe it should also be on first render?

Reply:
Using Milan's suggestion:
- No longer overriding `ShouldRender`
- Providing flag to guard when we call the interop code.